### PR TITLE
Support multiple STL files across same module

### PR DESCRIPTION
**Pull request by stopping**
* Support multiple STL files across same module

Modules can be split across multiple files. Example usage is reflected
in example/example.test, which imports example_base.stl (event
definitions, transition definitions, etc.) and example.stl (transition
instances).

* Allow STL constants to be defined in manifest

Constants can be declared in the STL without a value, and then defined
in the manifest via the 'constants' dictionary entry. Constants are
referenced in the manifest by '\<module\>::\<constant name\>'.

* Pass --manifest-args as a list instead of a string

If a manifest variable was a string with a space in it (for example,
--manifest-args="the_string='this has spaces in it'"), it could not be
parsed since we were splitting on whitespace. Now, we just pass the args
as a list (e.g. --manifest-args "this_string='has spaces'" "foo='bar'").

* Fixed quotes in README.md

Example code and commands should be copy/pastable directly from the
README.

### CODE REVIEW

File: README.md [hunk 1]
```diff
@@ -70,26 +69,35 @@ The **test manifest** describes the tests to be run. The test manifest file typi
...
+  'constants': {
+    'example::kHelloWorld': $hello_world,  # Must pass --manifest-args \
+                                           # 'hello_world="Hello, world!"'
+  },
+
+  'test': ['example::rReceiver'],
 }
 
-$ python test_driver.py example/example.test --manifest-args=”ip=’0.0.0.0’”
+$ python test_driver.py example/example.test --manifest-args 'ip="0.0.0.0"' \
```
Comment (protobird-git) [inline README.md]: $ python test_driver.py example/example.test \
       --manifest-args 'ip="0.0.0.0"' 'hello_world="Hello, world!"'

File: README.md [hunk 2]
```diff
@@ -54,11 +54,10 @@ positional arguments:
...
   -a MANIFEST_ARGS, --manifest-args MANIFEST_ARGS
-                        A series of space separated key=value pairs. Each
-                        instance of $key in the manifest file is replaced by
-                        value verbatim. In particular, if you want to pass a
-                        string, it must be explicitly quoted, e.g.:
-                        ip="0.0.0.0"
+                        A list of key=value pairs. Each instance of '$<key>' in
+                        the manifest file is replaced by <value> verbatim. In
+                        particular, if you want to pass a string, it must be
+                        explicitly quoted, e.g.: 'ip="0.0.0.0"'
```
Comment (protobird-git) [inline README.md]: Why did you change this?

File: README.md [hunk 3]
```diff
@@ -115,11 +123,32 @@ The ‘role’ key maps to the ‘module:rRoleName’, which in this case is ‘
...
 
-### 3.4. test
+### 3.4. constants
+The ‘constants’ field must be a dictionary. Each entry in the dictionary defines a declared constant described in the stl_files listed in the ‘stl_files’ field. In the above example, there [example_base.stl](example/example_base.stl) contains the following:
+```
+example_base.stl:
+
+module example;
+...
+const string kHelloWorld;
```
Comment (protobird-git) [inline README.md]: What is the initial value of this case? Or, does parser checks whether the constant is initialized somewhere?

File: README.md [hunk 4]
```diff
@@ -129,25 +158,27 @@ parameter_example.test:
...
-    FIELD=’new_field’ \
-    VALUE=’new_value’ \
+    FIELD=\'new_field\' \
+    VALUE=\'new_value\' \
     my_number=3 \
     my_boolean=True \
     str_no_quotes=World \
-    str_with_quotes=’bar’”
+    str_with_quotes=\'bar\' \
+    str_with_spaces="'foo bar baz'"
```
Comment (protobird-git) [inline README.md]: Hmm, what is the actual value in this case? with or without quotation?

File: README.md [hunk 5]
```diff
@@ -158,14 +189,15 @@ parameter_example_with_substitues.test:
...
         { 'role': 'example::rReceiver',
-          ‘new_field’: ‘new_value’,
-          ‘my_number’: 3,
-          ‘my_boolean’: True,
-          ‘str_no_quotes’: ‘Hello, World’,
-          ‘str_with_quotes’: ‘bar’,
+          'new_field': 'new_value',
+          'my_number': 3,
+          'my_boolean': True,
+          'str_no_quotes': 'Hello, World',
```
Comment (protobird-git) [inline README.md]: Something wrong in str_no_quotes.

File: example/example.test [hunk 6]
```diff
@@ -12,15 +12,26 @@
 # See the License for the specific language governing permissions and
 # limitations under the License.
 
+# This test can be run from this directory with:
+# PYTHONPATH=$PYTHONPATH:$PWD/.. \
+# python run_test.py example.test -a ip=\'127.0.0.1\' hello_world=\'Hello\'
```
Comment (protobird-git) [inline example/example.test]: why not hello_world="'Hello, world!'"?

File: stl/base.py [hunk 7]
```diff
@@ -413,16 +412,16 @@ def __str__(self):
 
   def __getitem__(self, key):
     if key not in self.fields:
-      raise AttributeError("No field exists in role '%s': %s" %
-                           (self.name, key))
+      raise AttributeError("No field exists in role '%s': %s" % (self.name,
```
Comment (protobird-git) [inline stl/base.py]: did you run gpylint? It doesn't look better though.

File: stl/base.py [hunk 8]
```diff
@@ -631,20 +630,19 @@ def __init__(self):
       self.test_source = False
 
     def __str__(self):
-      return ('CONTEXT: s(%s)%s, t(%s)%s' % (self.source, '*' if
-                                             self.test_source else '',
-                                             self.target, '*' if
-                                             not self.test_source else ''))
+      return ('CONTEXT: s(%s)%s, t(%s)%s' % (self.source, '*'
```
Comment (protobird-git) [inline stl/base.py]: better to break before '*'. Put () around '*' if self.test_source else ''?

File: stl/base.py [hunk 9]
```diff
@@ -631,20 +630,19 @@ def __init__(self):
...
+                                             if self.test_source else '',
+                                             self.target, '*'
```
Comment (protobird-git) [inline stl/base.py]: ditto

File: stl/module.py [hunk 10]
```diff
@@ -0,0 +1,64 @@
...
+    self.messages = {}
+    self.events = {}
+    self.transitions = {}
+
+  def __eq__(self, other):
+    return (isinstance(other, Module) and self.consts == other.consts and
+            self.roles == other.roles and self.states == other.states and
+            self.qualifiers == other.qualifiers and
+            self.messages == other.messages and self.events == other.events and
+            self.transitions == other.transitions)
```
Comment (protobird-git) [inline stl/module.py]: Don't check name? If not, please comment that why not.

File: test_driver.py [hunk 11]
```diff
@@ -102,45 +81,52 @@ def LoadManifest(manifest_filename, manifest_args):
...
 
 
+def FillInConstants(modules, manifest):
+  """Fills in constant information in |modules|."""
+  for key, val in manifest['constants'].iteritems():
+    module, name = key.split('::', 1)
+    if name not in modules[module].consts:
+      raise NameError("Cannot find a constant in module '%s': %s" % (module,
+                                                                     name))
+    modules[module].consts[name].value = val
```
Comment (protobird-git) [inline test_driver.py]: Please check if value is already set. Print warning if we want to allow overriding, otherwise, raise Error.

File: test_driver.py [hunk 12]
```diff
@@ -271,13 +256,14 @@ def Main():
...
-  modules = LoadModules(manifest, args.manifest, parse_env)
+  global_env = {}
+  modules = LoadModules(manifest, args.manifest, global_env)
 
-  if 'error' in parse_env and parse_env['error']:
+  if 'error' in global_env and global_env['error']:
     return False
 
   FillInModuleRoles(modules, manifest)
+  FillInConstants(modules, manifest)
```
Comment (protobird-git) [inline test_driver.py]: Please find uninitialized consts and raise Error if any.

File: README.md [hunk 13]
```diff
@@ -70,26 +69,35 @@ The **test manifest** describes the tests to be run. The test manifest file typi
...
 ```
 
 ### 3.2. stl_files
-The ‘stl_files’ field must be a list of strings naming the STL files that include the transitions to be tested. The filenames here should be relative to the manifest .test file. In the example above, [example.test](example/example.test) and [example.stl](example/example.stl) are in the same directory.
+The ‘stl_files’ field must be a list of strings naming the STL files that include the transitions to be tested. The filenames here should be relative to the manifest .test file. In the example above, [example.test](example/example.test) and [example_base.stl](example/example_base.stl) are in the same directory.
 
 ### 3.3. roles
-The ‘roles’ field must be list of dicts. Each dict element in the list describes one role. These are the roles described in the stl_files listed in the ‘stl_files’ field. In the above example, there [example.stl](example/example.stl) contains the following:
+The ‘roles’ field must be list of dicts. Each dict element in the list describes one role. These are the roles described in the stl_files listed in the ‘stl_files’ field. In the above example, there [example_base.stl](example/example_base.stl) contains the following:
```
Comment (mbjorge) [inline README.md]: nit: I know this was the original wording, but I think the 'there' should get removed so it just reads: In the above example, example_base.stl contains the following

File: README.md [hunk 14]
```diff
@@ -115,11 +123,32 @@ The ‘role’ key maps to the ‘module:rRoleName’, which in this case is ‘
 
 The remaining key: value pairs map to the fields of the role as described in the STL file.
 
-### 3.4. test
+### 3.4. constants
+The ‘constants’ field must be a dictionary. Each entry in the dictionary defines a declared constant described in the stl_files listed in the ‘stl_files’ field. In the above example, there [example_base.stl](example/example_base.stl) contains the following:
```
Comment (mbjorge) [inline README.md]: same here. Remove the 'there' before example_base.stl

File: README.md [hunk 15]
```diff
@@ -115,11 +123,32 @@ The ‘role’ key maps to the ‘module:rRoleName’, which in this case is ‘
...
+
+The key maps to the ‘<module>::<constantName>’, which in this case is ‘example::kHelloWorld’.
+
+### 3.5. test
 The ‘test’ field must be a list of strings naming the roles to be tested. Each entry in the test list must correspond to a ‘role’ value in the ‘roles’ list.
 
-### 3.5. Substituting Parameters
-Values can be substituted from the command line by passing a space separated list of key=value pairs to the --manifest-args option of test\_driver.py. The manifest .test file will then replace any occurence of $key with value. $key can be located anywhere in the manifest file (e.g. in the list of stl\_files, as a key or vparameter_example.test: test).
+### 3.6. Substituting Parameters
+Values can be substituted from the command line by passing a space separated list of key=value pairs to the --manifest-args option of test\_driver.py. The manifest .test file will then replace any occurence of $key with value. $key can be located anywhere in the manifest file (e.g. in the list of stl\_files, as a key or value in any of the manifest dictionaries).
```
Comment (mbjorge) [inline README.md]: nit: Above you used the notation $<key> and <value>, but here it's still $key and value. Either way is fine, just make sure it's consistent

File: README.md (hunk 3 above)
Comment (mbjorge) [inline README.md]: If you specify a value in the .stl file, does the manifest value override it, or does it throw an error?

File: README.md (hunk 1 above)
Comment (stopping) [inline README.md]: Done

File: stl/parser_test.py [hunk 16]
```diff
@@ -32,31 +33,21 @@ class StlParserTest(unittest.TestCase):
   TEST_FILENAME = 'dummy.stl'
 
   def setUp(self):
-    self.actual_module_dict = {}
-    stl.parser._InitializeModuleDict(self.actual_module_dict)
-    self.parse_env = {}
+    self.actual_module = stl.module.Module('foo')
+    self.parse_env = {'modules': {'foo': self.actual_module}}
```
Comment (mbjorge) [inline stl/parser_test.py]: global_env? here and below, to match the change from above

File: README.md (hunk 2 above)
Comment (stopping) [inline README.md]: We don't pass the entire manifest-args as a string anymore; instead we pass a list of strings so that we don't have to split on whitespace (which breaks when you try to use a manifest argument that has a space in it). I changed the comment to fit the new argument structure.

File: README.md (hunk 3 above)
Comment (stopping) [inline README.md]: It initializes to None, the test driver can easily check if all constants are initialized properly.

File: README.md (hunk 13 above)
Comment (stopping) [inline README.md]: Done

File: README.md (hunk 14 above)
Comment (stopping) [inline README.md]: Done

File: README.md (hunk 3 above)
Comment (stopping) [inline README.md]: Good point, it should probably throw an error. Done.

File: README.md (hunk 15 above)
Comment (stopping) [inline README.md]: Done.

File: README.md (hunk 4 above)
Comment (stopping) [inline README.md]: The shell will evaluate the argument as the following:

str_with_spaces='foo bar baz'

You could wrap the whole argument in quotes as well, as:
"str_with_spaces='foo bar baz'"

We can't escape the inner single quotes since the spaces will separate on the command line.

File: README.md (hunk 5 above)
Comment (stopping) [inline README.md]: Done.

File: example/example.test (hunk 6 above)
Comment (stopping) [inline example/example.test]: Done.

File: stl/base.py (hunk 7 above)
Comment (stopping) [inline stl/base.py]: I did. I'm okay with the slightly weird formatting; it's better than correcting gpylint every time we want to run it I think.

File: stl/base.py (hunk 8 above)
Comment (stopping) [inline stl/base.py]: Done

File: stl/base.py (hunk 9 above)
Comment (stopping) [inline stl/base.py]: Done

File: stl/module.py (hunk 10 above)
Comment (stopping) [inline stl/module.py]: Done.

File: stl/parser_test.py (hunk 16 above)
Comment (stopping) [inline stl/parser_test.py]: Done.

File: test_driver.py (hunk 11 above)
Comment (stopping) [inline test_driver.py]: Done.

File: test_driver.py (hunk 12 above)
Comment (stopping) [inline test_driver.py]: Done.

File: README.md (hunk 4 above)
Comment (stopping) [inline README.md]: Actually, we need to use single quotes around the whole argument, and double quotes around the value due to the way bash evaluates strings. Fixed this in the doc.

File: README.md [hunk 17]
```diff
@@ -158,14 +189,15 @@ parameter_example_with_substitues.test:
...
-          ‘my_number’: 3,
-          ‘my_boolean’: True,
-          ‘str_no_quotes’: ‘Hello, World’,
-          ‘str_with_quotes’: ‘bar’,
+          'new_field': 'new_value',
+          'my_number': 3,
+          'my_boolean': True,
+          'str_no_quotes': 'Hello, World',
+          'str_with_quotes': 'bar',
+          'str_with_spaces': "foo bar baz",
```
Comment (protobird-git) [inline README.md]: 'foo bar baz' instead of " since python prefers consistency in quotation marks?

Comment (stopping) [inline README.md]: Done.
